### PR TITLE
RFC: Introducing AltNet, a pluggable framework for alternative transports

### PR DESCRIPTION
**Pull request by ariard**
**As of 14/03/2022 (chain tip hash : 00000000000000000000628002621832be9281faf72c394e66a8bc8446f80c75), this project is under implementation in separate process relying on the multiprocess framework. See https://github.com/chaincodelabs/libmultiprocess/issues/56 for more information on current state.**

This PR proposes to introduce a new subsystem aiming towards providing a generic framework for integrating alternative transport communication channels inside Bitcoin Core. This PR is a functional proof-of-concept using LN transport protocol as a new link layer for header fetching. See https://github.com/bitcoin/bitcoin/issues/18989 for high-level discussion and proposal merits.

## The Design 

Previous [work](https://github.com/bitcoin/bitcoin/pull/17376) has been attempted to increase network security and robustness by increasing communication redundancy. While being optional and fault-tolerant, these attempt don't scale well to integrate a wide-range of alternatives such as [domain-fronting](https://www.bamsoftware.com/papers/fronting/), [orbital communications](https://github.com/Blockstream/satellite), [amateur radio](https://en.wikipedia.org/wiki/DASH7), [Tor pluggable transports](https://snowflake.torproject.org/), unlicensed UHF, Bluetooth, [I2P](https://geti2p.net/en/), Lightning[ Noise transport](http://noiseprotocol.org/), ... Each of these transports has its own availability, latency, don't bind to the Berkeley sockets model, are unidirectional, require custom peer discovery, are not suited for block traffic, ...

This proposal tries to overcome these issues by laying out a generic "driver" framework with its own new processing stack, thereby encapsulating per-transport oddities. It introduces:

* Per-transport capabilities (`TransportCapabilities`). Originally declared by each driver, they abstract driver specifities to general flags, like directionality, protocol messages scope. They could be extended with further attributes, like max bandwidth for adjusted packing or "privacy"/"trustness" niceness of the link.
* Abstract driver model (`CDriver`). Basically, it dissociate driver warmup, message preprocessing, sending and reception. Ideally you would want to load them dynamically.
* "Anycast" header-fetching (`AltLogicValidation::BlockHeaderAnomalie`). After notification from the [watchdog module](https://github.com/bitcoin/bitcoin/pull/18987), it sends a `GETHEADERS` message to any peers supporting bidirectionality and headers processing.
* Alternative processing stack (`AltLogicValidation`). Relying on its own threads, it manages drivers and may take action based on static logical capabilites instead of peer state, reducing likelihood of bugs. In the future, one could implement a more robust headers-syncing strategy like reverse-header or compressed headers. It may also leverage a [future asynchronous validation interface](https://github.com/bitcoin/bitcoin/pull/18963).

There are many advantages of this kind of design --, it is easy to integrate new drivers binding to an abstract class, it is easy to reason about due to logical capabilities and at the same time provides granular tuning due to the fine-integration with rpc. Peer discovery and policy or bandwidth/memory DoS-protection are all possible driver add-ons as well. One could envision an rpc like `sendemergencytx`, broadcasting a transaction on any transport protocol available, like a [LN's justice transaction](https://github.com/lightningnetwork/lightning-rfc/blob/master/05-onchain.md#revoked-transaction-close-handling) after detection of being eclipsed.

This module could be made entirely optional at compile-time with `--enable-altnet` even if such integration with the build system hasn't been made.

Obviously this is an exploratory PoC, I don't expect code here to be merged. Beyond worthiness of doing this in Rust there are a lot of open questions, including the threading/memory model, interface with existing codebase, build system, resources consumption and fault-tolerance. As next steps, I would favor building on the multiprocess work and hosting this new network stack in its own to avoid any bug contamination from introducing complex driver. I encourage those with opinions on the higher level motivation and design discussion to leave their ideas on the linked issue.

High-level design:

              ___________            ______________________
             |           |          |                      |
             | CWatchdog |----------|  AltLogicValidation  |
             |___________|          |______________________|
                                               |
                                         ______|______
                                        |             |
                                        |  CAltstack  |                                                         
                                        |_____________| 
                                               |
                                           ____|____
                                          |         |
                                          | CDriver |
                                          |_________|
                                               |
                                       ________|__________
                                      |                  |
                                      | ClightningDriver |
                                      |__________________|
                                      
                                                   

## The Demo

This PoC features a header-fetching round-trip, leveraging LN connection protocol. Each LN node connects to its peers through an encrypted and authenticated channel. Since your LN peer must have chain access, you could leverage this privileged link to create a redundant chain access.

Using C-lightning [custom LN message support](https://github.com/ElementsProject/lightning/pull/3315) and [plugin API](https://github.com/ElementsProject/lightning/blob/master/doc/PLUGINS.md), we could embed directly inside `HEADERS`/`GETHEADERS` and realize a round-trip between 2 disconnected bitcoin nodes through their respective C-lightning nodes.

For this to work, you need to compile and run 2 C-lightning nodes,  each with its own data directory and launched with this [custom bridge plugin](https://github.com/ariard/bridge-layer). Once those LN nodes are running, activate bridges by `lightning-cli alice_on` and `lightning-cli bob_on`. The plugin doesn't seem to accept CLI argument so I had to hardcode some node addresses.

Then compile 2 bitcoin nodes with this patchset. For one of them you must change the hardcoded port in C-lightning driver and deactivate the watchdog logic. Also, it would be best if you changed `SCAN_INTERVAL` to something reasonable, for instance 1 min.

High-level setup:

                                                 -->

                    __________          LN_msg(`GETHEADERS`)                   ________
                   |          |                                               |        |
                   | CL-Alice |-----------------------------------------------| CL-Bob |
                   |__________|                                               |________|           
                        |               LN_msg(`HEADERS`)                         |
                    ____|_____                                                 ___|_____
                   |          |                 <--                           |         |
                   |  bridge  |                                               | bridge  |
                   |__________|                                               |_________|
                        |                                                         |
                    ____|_____                                                 ___|____
                   |          |                                               |        |
                   | BC-Alice |                                               | BC-Bob |
                   |__________|                                               |________|

                      HEIGHT                                                   HEIGH + 1

Running both nodes, they should automatically connect to the bridge, and after anomaly detection realize a header exchange, if one of them is at least one block forward (run `generatetoaddress 1 $ADDR`).

Otherwise you may observe past logs:

Alice:
```
Ê0-05-16T00:55:11Z Boostrapping Altstack
2020-05-16T00:55:11Z init message: Done loading
2020-05-16T00:55:11Z msghand thread start
2020-05-16T00:55:11Z drivers-warmup thread start
2020-05-16T00:55:11Z addcon thread start
2020-05-16T00:55:11Z Registered driver with 0
2020-05-16T00:55:11Z net thread start
2020-05-16T00:55:11Z altstack-processing thread start
2020-05-16T00:55:11Z Clightning - Warmup
2020-05-16T00:55:11Z dnsseed thread start
2020-05-16T00:55:11Z altstack-handle thread start
2020-05-16T00:55:11Z 0 addresses found from DNS seeds
2020-05-16T00:55:11Z dnsseed thread exit
2020-05-16T00:55:11Z Clightning - Bridge connection
2020-05-16T00:55:11Z Registering node_id 0 with driver_id 0
2020-05-16T00:55:11Z drivers-warmup thread exit
2020-05-16T00:55:16Z Block header anomalie detected - notifying subscribers
2020-05-16T00:55:16Z Block header anomalie detected - Anycasting header fetching from 0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206
2020-05-16T00:55:16Z Clightning - Preprocessing getheader
2020-05-16T00:55:16Z Cligthning - Bridge send success 71
2020-05-16T00:55:19Z Cligtning - Bridge receive success
2020-05-16T00:55:19Z Receive HEADERS from 0
2020-05-16T00:55:19Z Receiver header 65084efeaffe9929de0a0bb747cf48ac7f9b4edb214b2d1ad2ea468977be8e1e vs tip 0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206
```

Bob:
``` 
2020-05-16T00:55:09Z Boostrapping Altstack
2020-05-16T00:55:09Z addcon thread start
2020-05-16T00:55:09Z opencon thread start
2020-05-16T00:55:09Z msghand thread start
2020-05-16T00:55:09Z drivers-warmup thread start
2020-05-16T00:55:09Z Registered driver with 0
2020-05-16T00:55:09Z Clightning - Warmup
2020-05-16T00:55:09Z init message: Done loading
2020-05-16T00:55:09Z dnsseed thread exit
2020-05-16T00:55:09Z altstack-processing thread start
2020-05-16T00:55:09Z altstack-handle thread start
2020-05-16T00:55:09Z Clightning - Bridge connection
2020-05-16T00:55:09Z Registering node_id 0 with driver_id 0
2020-05-16T00:55:09Z drivers-warmup thread exit
2020-05-16T00:55:10Z Leaving InitialBlockDownload (latching to false)
2020-05-16T00:55:17Z Cligtning - Bridge receive success
2020-05-16T00:55:17Z Receive GETHEADERS from 0
2020-05-16T00:55:17Z Fetch header starting from 0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206, tip 65084efeaffe9929de0a0bb747cf48ac7f9b4edb214b2d1ad2ea468977be8e1e
2020-05-16T00:55:17Z Sending back 1 headers
2020-05-16T00:55:17Z Clightning - Preprocessing header
2020-05-16T00:55:17Z Cligthning - Bridge send success 84
2020-05-16T00:56:10Z Adding fixed seed nodes as DNS doesn't seem to be available.
```

### CODE REVIEW

**Comment by icota** [comment]
Exciting stuff but I question the logic of there being a framework just for "alternatives". Shouldn't ipv4/6 and onion be one of many drivers with their own capabilities and trustworthiness?

More work initially but probably easier maintenance down the road and a guarantee that any type of  smoke signal protocol driver could be a first-class citizen like ip.

Concept NACK for not being ambitious enough :wink:

File: src/altnet.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_ALTNET_H
+#define BITCOIN_ALTNET_H
+
+#include <alt_processing.h>
+#include <drivers.h>
+#include <net.h>
+#include <netmessagemaker.h>
```
Comment (TheBlueMatt) [inline src/altnet.h:12]: I think this should be way more disconnected from the existing P2P stack. Accessing anything but validation.h and some utils seems bad.

**Review by TheBlueMatt — COMMENTED** [review]
This is really cool! Its certainly the direction I was playing around with with the attempts at adding Rust stuff. That said, I'm really not sure the API is right, though - it seems much too tied to the existing serialization/network framework, which both misses a good chunk of the redundancy possibility and prevents things like blocks-over-HTTP(s), DNS, radio, etc, etc.

**Comment by ariard** [comment]
@TheBlueMatt what's about a thread pool for every blocking alternative and a `CDriver::IsReady` when we try to read from it ? Have you in mind other alternative which needs a really specific threading model ? That's a bit more complexity but you may allocate priority to avoid starvation.

I agree that the `mapNodeState` isn't great and ideally you do want to defer every peer policy to the driver. Reducing state is less error-prone.

@icota Do you have in mind a different ipv4/ipv6 flow that the one implemented by main net ? Path-aware routing could be cool and definitely something belonging there.

**Comment by willcl-ark** [comment]
Hello @ariard pluggable alternative transport systems for Bitcoin Core gets a concept ACK from me.

I wonder if in the future this framework could include some transport prioritisation which might, for example, relay blocks over an internet/Tor connection but attempt to broadcast and relay transactions over Amateur radio/bluetooth (as available) with a new Tor circuit as a fallback, to further obscure a transactions origin.

In the case that your node might have a metered main connection e.g. 3G along with a secondary transport mechanism, the same prioritisation might be able to help split bandwidth required between multiple transports simultaneously.

For the low bandwidth/metered scenario, I feel this would be an excellent match for the compressed block headers scheme I picked up on in my email [here](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-May/017834.html).

**Comment by remyers** [comment]
I love this concept! I will need some more time to have implementation comments, but will try to think about this in the context of the alt net I'm most familiar with (stateless unlicensed UHF Mesh / goTenna). Some preliminary big picture thoughts:

 1) should the Watchdog functionality eventually be broken out as its own RFC? even without an altnet transport wouldn't a core wallet also want notification of unusual p2p events?

 2) Right now core is aware of [ASN buckets](https://bitcoincore.reviews/16702.html), do think different altnet connections could be treated as diverse ASNs?
 2a) Should altnet peers be treated as one ASN, or is there a way to ensure diversity to protect against sybil altnet peers.

 3) Do you envision a plugin-driver model so that core would support the high level CDriver/Watchdog, but people could plug-in their own low-level implementations? 
 3a) In the case of goTenna, we have a low-level C driver that is not open source so would not be appropriate to include in core. Would something like the HWI system make sense for different vender specfic CDriver instances?

 4) A deeper integration to make altnet peers 1st class peers would be difficult, but if a CNetAddr worked more like a URI then peers could be stored as ip:\\ipaddress:port and altnet nodes would look like clightning:\\peeraddr and use an altnet CDriver named `clightning`.
 4a) A less invasive approach to 1st class would be that any altnet peer just uses 127.0.0.1:cdriver_port, but all peers can include user configurable transport capabilities.

Great work, I'm really excited to try this out.

**Comment by MarcoFalke** [comment]
Concept ACK, but I am wondering if some of this would be better off outside the Bitcoin Core build system. Obviously stuff that can be done with no additional requirements (additional requirements could be a proxy, another daemon or even new hardware like an antenna) and that is a clear benefit to Bitcoin Core can live inside of Bitcoin Core. For example something like the existing `TipMayBeStale` logic.

However, for transports that would pull in additional dependencies, a more flexible approach would be to define a common interface. The drivers could then ask Bitcoin Core for raw validation data and statistics, then reply with new headers or blocks. For example, this could be done over the RPC interface. I haven't thought about all the trade-offs, but it seems beneficial if addons can be attached and removed during run-time as well as being developed independently of the Bitcoin Core development process?

**Comment by ariard** [comment]
Hi @willcl-ark,

> I wonder if in the future this framework could include some transport prioritisation which might, for example, relay blocks over an internet/Tor connection but attempt to broadcast and relay transactions over Amateur radio/bluetooth (as available) with a new Tor circuit as a fallback, to further obscure a transactions origin.

Yes introducing drivers capabilities (`fHeaders`, `fTransaction`, ...) should be seen as primitives to label traffic according to application/user requirements. I can foresee `bitcoin-cli ioctl bluetooth --label=tx,headers`, you're envisioning something like this ?

> In the case that your node might have a metered main connection e.g. 3G along with a secondary transport mechanism, the same prioritisation might be able to help split bandwidth required between multiple transports simultaneously.

Ah thanks didn't think about this, effectively when bandwidth has been consumed on one device you may want to fallback on another one. I do think bandwidth tracking should be done drivers-side.

> For the low bandwidth/metered scenario, I feel this would be an excellent match for the compressed block headers scheme I picked up on in my email here.

Yes IIRC compressed block headers in the linked issue (#18989). Clearly that would be easier to innovate on message protocol, like also transaction compression.

**Comment by ariard** [comment]
@remyers 

> should the Watchdog functionality eventually be broken out Do you envision a plugin-driver model so that core would support the high level CDriver/Watchdog, but people could plug-in their own low-level implementations?
3a) In the case of goTenna, we have a low-level C driver that is not open source so would not be appropriate to include in core. Would something like the HWI system make sense for different vender specfic CDriver instances?as its own RFC? even without an altnet transport wouldn't a core wallet also want notification of unusual p2p events?

Yes Watchdog has its own PR (#18987), with a `CWatchdogInterface`. It could be used by any kind of application, including the wallet.

> Right now core is aware of ASN buckets, do think different altnet connections could be treated as diverse ASNs?
2a) Should altnet peers be treated as one ASN, or is there a way to ensure diversity to protect against sybil altnet peers.

I would like to avoid dealing with ASN/IP address at all. If you assume wireless communications there is no ASN to deal with, peer discovery and policy should be inside the driver. With regards to peer, you may only need a general way to authenticate them, so maybe some driver method enumerating its hosted peers ?

> Do you envision a plugin-driver model so that core would support the high level CDriver/Watchdog, but people could plug-in their own low-level implementations?
3a) In the case of goTenna, we have a low-level C driver that is not open source so would not be appropriate to include in core. Would something like the HWI system make sense for different vender specfic CDriver instances?

I clearly envision a plugin-driver model and exactly core would support the high-level `CDriver/`CWatchdog`/`AtLogicValidation`. I lean to something like HWI and drivers implementations being hosted outside of core, under bitcoin-core/ for the open-sources ones. With regards to proprietary, I don't have opinion, if they bind to the interface that's better for you I guess.

> A deeper integration to make altnet peers 1st class peers would be difficult, but if a CNetAddr worked more like a URI then peers could be stored as ip:\ipaddress:port and altnet nodes would look like clightning:\peeraddr and use an altnet CDriver named clightning.
4a) A less invasive approach to 1st class would be that any altnet peer just uses 127.0.0.1:cdriver_port, but all peers can include user configurable transport capabilities.

I don't want to interfere with the actual p2p stack. But if what you want to address is user configurable transport capabilities that can be done at the rpc-level and yes something like `bitcoin-cli ioctl Wifi:XXX --label=tx,addr` ?

Thanks for the feedback!

**Comment by ariard** [comment]
> However, for transports that would pull in additional dependencies, a more flexible approach would be to define a common interface.

Yes that's the idea defining a `CDriver` interface, even if it needs refinement. `drivers/` should be elsewhere and you just keep the fetching logic/interface in Core. I don't think drivers should ask for raw validation data on their own, i.e you may receive your block from actual p2p but want to broadcast it over WiFi. Drivers acting on their own means you poll a lot of nothing due to block variance, and that's what RPC doesn't fit IMO. Haven't thought that much about run-time addition but I completely agree to not encumber Bitcoin Core development process more. You provide a generic framework and drivers devs can work on their own in another repo.

**Comment by TheBlueMatt** [comment]
> but I am wondering if some of this would be better off outside the Bitcoin Core build system.

@MarcoFalke sadly, such a design kicks something out to live in purgatory forever. We've never materially seen any real adoption of "additional, on the side" contrib things. Even opt-in configuration is rarely used, let alone something with additional setup requirements. I don't think this carries much value unless its targeting (eventually) being built-in-by-default.

**Comment by TheBlueMatt** [comment]
> @TheBlueMatt what's about a thread pool for every blocking alternative and a CDriver::IsReady when we try to read from it ? Have you in mind other alternative which needs a really specific threading model ? That's a bit more complexity but you may allocate priority to avoid starvation.

I don't know that the framework even needs to provide any kind of threading? Why not let the applications decide that. Using the net.h stuff like NetMsgType, CBlockLocator, NetMsgMaker and the connection types really limits the utility here. eg radio stuff probably wants to read/write from a tty, not a "real socket".

**Comment by ryanofsky** [comment]
I'm starting to look at this and my first impressions are basically the things Matt's saying: that the API for transports can probably be very minimal and not deal with threading, and not interact with (or duplicate and fork) existing net_processing and net code.

A real anomaly detection implementation, on the other hand, may need to have access to more things, but hopefully that could happen more by passively receiving events than actively looking up state.

I think the writeups in #18987, #18988, #18989 provide a good description of the features this is trying to support, and the PR code provides a good proof or concept and sketch of design. But one thing that's missing and would be nice would be a plan for how this might be rolled out. I'm assuming we wouldn't add a new framework before anything was using it, and I know Matt has already implemented a lot of things that could use this framework. So it'd be great if there was a writeup (or outline like the [assumeutxo plan](https://github.com/jamesob/assumeutxo-docs/tree/2019-04-proposal/proposal#plan)) saying what the steps could be for rolling this out, and linking to Matt's implementations, and saying how things could be integrated. The [dev wiki](https://github.com/bitcoin-core/bitcoin-devwiki) could be a good place to put this, since it would be speculative and subject to change.

On the PR itself, I know it is exploratory and you don't expect it to be merged, but was hard not to notice it was copying and pasting a lot of existing code from net/net_processing/validationinteface code and also introducing new code following older conventions (`CPrefixedClassNames` `bHungarianVariables`), so I just want to quickly encourage using the current [style](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c) in new code and trying to avoid duplicate implementations of things like [`WatchSignalsInstance`](https://github.com/ariard/bitcoin/blob/3667d18a76f72acf0741a886715b6c9fee40001d/src/watchdoginterface.cpp#L14) / [`MainSignalsInstance`](https://github.com/ariard/bitcoin/blob/3667d18a76f72acf0741a886715b6c9fee40001d/src/validationinterface.cpp#L26) in the real version of this.

**Comment by ariard** [comment]
> Using the net.h stuff like NetMsgType, CBlockLocator, NetMsgMaker and the connection types really limits the utility here. eg radio stuff probably wants to read/write from a tty, not a "real socket". Let alone something like DNS which doesn't even have a file descriptor, only a libc call to poll.

I think message translation from a wire format suiting communication channel should be done by the driver, it doesn't make sense at all to make it in the framework. So we may pass `NetMsgType` to the driver interface without caring further.

With regards to threading right, but isn't the case in main stack we have a separation between `ThreadSocketHandler` and `ThreadMessageHandler` for performance reasons ? We may follow there ?

**Comment by ariard** [comment]
Thanks @ryanofsky for review !

> and not interact with (or duplicate and fork) existing net_processing and net code.

Right, what are all the reasons between net_processing and net code separation ?

> A real anomaly detection implementation, on the other hand, may need to have access to more things, but hopefully that could happen more by passively receiving events than actively looking up state.

Right, `CWatchdog` may be consumer of an asynchronous interface too, I keep it simple for a PoC but that's a good suggestion.

>  So it'd be great if there was a writeup (or outline like the assumeutxo plan) saying what the steps could be for rolling this out

Yes it's here : https://github.com/ariard/altnet-proposals/tree/master, still WIP. Note: I've been talking with folks actively working on alt-com like BlockSat. As first functional step , I plan to reuse the headers-over-DNS from Matt, it's pretty straightforward and can be used by anyone without further requirement.

> On the PR itself, I know it is exploratory and you don't expect it to be merged, but was hard not to notice it was copying and pasting a lot of existing code from net/net_processing/validationinteface

Right, it may make sense to refactor `MainSignalsInstance` to make it generic enough for reuse.

**Comment by luisan00** [comment]
> Concept ACK, but I am wondering if some of this would be better off outside the Bitcoin Core build system. ...

Totally agree.
In principle could be better as a separate project. It seems interesting but as you say, a common interface can help in this particular, separating transport logic from the Bitcoin Core logic.

I like the idea and will think carefully the next days in the context that is most familiar to me, ISM and low power mesh networks, and in particular Locha Mesh.

**Comment by icota** [comment]
> Do you have in mind a different ipv4/ipv6 flow that the one implemented by main net ? Path-aware routing could be cool and definitely something belonging there.

My point is that rather than making a separate framework I would prefer to change `net` to support the functionality you are suggesting for AltNet. 

Look at what I am greeted with when I run `getnetworkinfo`:
```
"networks": [
    {
      "name": "ipv4",
      "limited": false,
      "reachable": true,
      "proxy": "",
      "proxy_randomize_credentials": false
    },
    {
      "name": "ipv6",
      "limited": false,
      "reachable": true,
      "proxy": "",
      "proxy_randomize_credentials": false
    },
    {
      "name": "onion",
      "limited": true,
      "reachable": false,
      "proxy": "",
      "proxy_randomize_credentials": false
    }
  ],
```
Right now this output is produced by iterating over a dumb enum but would it not be cool if you could plug a "satellite" or "radio" driver in there. 

This approach to alternative transports seems very intuitive to me but I don't see the sentiment echoed by anyone else. I haven't dug deep so it makes me think it's either not a good idea to begin with and/or involves some very heavy lifting?

Have you considered it?

**Comment by ariard** [comment]
@icota 

> My point is that rather than making a separate framework I would prefer to change net to support the functionality you are suggesting for AltNet.

I spend some time thinking to integrate this in our currently net/net_processing.

With regards to net you need to abstract any socket operation behind some interface, make peer stateful to know which stack they are relying on and threading may not fit, cf Matt argument on polling.

With regards net_processing, we currently assume bidirectionality, one-transport-fit-all (melding our own tx announcement with public block-relay) and timing based on always existing connection.

Now, considering the user interface it should be unified, you may host both altnet and net behind same commands, i.e in your example "radio" and "blocksat" would present too, just with a field `altnet=true`. Ideally traffic assignement ("this peer will do block-relay") should be cross-stack

**Comment by Ayms** [comment]
I am not expert at all with bitcoin code so I don't know what approach could be the best

But let's take another network example, IPFS, despite of their efforts (ie libp2p) it is not so simple to plug "other codebase" to the system, see  ``https://github.com/ipfs/ipfs/issues/439#issuecomment-593116241``, even with the same techno

And as we can see the codebase I am referring too here is https://github.com/Ayms/node-Tor

The plan right now for bitcoin would be ``bitcoin | node-Tor | bitcoin``, with (taking @icota approach):

	"networks": [
	  ...
		{
		  "name": "unix pipe",
		  ...
		},
  

Yes it's a bit ugly but I don't see other way today without using a localhost/socks, which I don't want, even if suppported, for sure it would be better to pipe directly to the nodejs module (is there a way to do unix pipe today for some potential testing ?)

And more ambitious at the end would be to be able to plug also to something like http://www.peersm.com/Convergence-2020.pdf "A universal and generic architecture to anonymize any application or protocol and turn it into an independent decentralized p2p network inside browsers and servers, with browsers acting as servers", including WebRTC, browsers, etc

**Comment by ryanofsky** [comment]
re: https://github.com/bitcoin/bitcoin/pull/18988#issuecomment-642561533

> Yes it's a bit ugly but I don't see other way today without using a localhost/socks, which I don't want, even if suppported, for sure it would be better to pipe directly

Not completely sure about the context but 52192725a3ec1f814cf40049de45efffc226cd5b from #19160 is an example showing how to spawn a new process and communicate with it bidirectionally over a pipe. It requires a some boilerplate to define a new [`src/interfaces/` class](https://github.com/bitcoin/bitcoin/blob/52192725a3ec1f814cf40049de45efffc226cd5b/src/interfaces/echo.h#L14-L22) and [`interfaces::Init::make` method](https://github.com/bitcoin/bitcoin/blob/52192725a3ec1f814cf40049de45efffc226cd5b/src/interfaces/init.h#L27), but after you do that it's easy to add methods to the interface, and make method calls from either process to the other with call forwarding and serialization and threading handled automatically. 

Additionally with 01c31f9dcd15c1009c4b69e820fc0c01b73cd3c9 from the [Process Separation](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Process-Separation) branch [ryanofsky@`ipc-export`](https://github.com/ryanofsky/bitcoin/commits/ipc-export), bitcoin listens on a `<datadir>/sockets/node.sock` socket that other processes can connect to and call the same interfaces.

**Comment by icota** [comment]
I almost forgot since it was some time ago but in c-lightning we added local socket (`AF_UNIX`) support (https://github.com/ElementsProject/lightning/pull/1323, https://github.com/ElementsProject/lightning/pull/1450) and with a couple of processes in between I ran the protocol over NFC.

Pipes and sockets approach can support a lot (but not all) of "alternative transport" use cases and seems like low hanging fruit. It is definitely something to consider.

**Comment by ariard** [comment]
@icota 

> Pipes and sockets approach can support a lot (but not all) of "alternative transport" use cases and seems like low hanging fruit. It is definitely something to consider.

After reviewing Russ PR (#19160), you can reuse process spawning and ipc framework to let communicate between a `bitcoin-node` and `bitcoin-altnet` one. And then again it could be reused to let `bitcoin-altnet` talks with any driver, each its own process. These drivers may not use a socket approach for the alternative transport they are supporting.

Pipes and sockets seems to me suited for "internal" communications between our collections of processes, not mandatory for talking with external, or do you see things differently ?

**Comment by ariard** [comment]
@Ayms 

> Yes it's a bit ugly but I don't see other way today without using a localhost/socks, which I don't want, even if suppported, for sure it would be better to pipe directly to the nodejs module (is there a way to do unix pipe today for some potential testing ?)

What issue do you have with sockets ? Let's envision this setup : `bitcoin-node <-> tor-driver <-> node-tor` where `<->` denotes Unix socket as IPC mechanism.

Using pipes you need process to be related and share the same file descriptor table which means integrating node-tor code to bitcoin one which isn't considered. Unless I miss some unix tricks to pass pipe handles across process ?

**Comment by Ayms** [comment]
@ariard , no real issue with sockets and (as you write) it should better be ``bitcoin <-> node-Tor`` and not ``bitcoin -> node-Tor -> bitcoin`` (basic stdin/stdout), I was talking about unix sockets in my previous message (which I assimilated to unix pipe in terms of wording) and not inter unix processes pipes 

I would just modify ``bitcoin-node <-> bitcoin-altnet <-> tor-driver <-> node-tor`` by ``bitcoin <-> node-Tor (via altnet)`` where node-Tor implements IPC or (better) ``bitcoin-node <-> altnet nodejs-driver <-> node-tor``, so nodejs modules don't have to implement unix sockets

Where ``node-js-driver`` is something like:

	altnet.on(‘data’,//process data from unix socket)
	altnet.send=//send data to unix socket
	
Then nodejs programs would just import altnet and turn it into the required object (a [Duplex Object](https://github.com/Ayms/node-Tor/blob/master/docs/README.md#the-duplex-object-and-evented-pipes) for node-Tor), two lines of code and that's it (unfortunately as mentioned above this does not work very well with some existing examples (ipfs), but for me this is a misdesign of ipfs that should not be reproduced)

Indeed, in @icota example "networks" could be turned into "layers" and probably a good idea would be to be able to pipe between altnet layers when applicable (``bitcoin-node <-> altnet some driver <-> my_super_encryption <-> altnet nodejs-driver <-> node-tor``, it looks similar to what @icota proposes ``core ---> relay process ---> (some sort of alternative transport) ---> relay ---> other core``

No invention here, this is what unix does since the beginning of times, and similar to #9979 attempt, probably some code already exists to do this (ie easily plug layers from one techno/process to another)
	
@ryanofsky , thanks for all the links, I will look at it in more details when I need it but it still looks a bit complicate, for simple testing, is there a simple way (modif in the bitcoin core code) to redirect/listen the bitcoin protocol via stdin/stdout?

**Comment by ryanofsky** [comment]
> @ryanofsky , thanks for all the links, I will look at it in more details when I need it but it still looks a bit complicate, for simple testing, is there a simple way (modif in the bitcoin core code) to redirect/listen the bitcoin protocol via stdin/stdout?

With #10102 you could start unmodified bitcoin-node or bitcoin-wallet binaries via [socat](http://www.dest-unreach.org/socat/) if you wanted to control them with the ipc protocol over stdin/stdout:

```sh
socat - system:"src/bitcoin-node -ipcfd 4",fdin=4,fdout=4
```

I could actually tweak the multiprocess spawning code to communicate over stdin/stdout instead of a separate socketpair descriptor, so you wouldn't need socat. But one reason the implementation avoids stdin/stdout is that it makes printf debugging awkward, and one stray printf or `-printtoconsole` anywhere in the codebase messes up the IPC connection.

Also as mentioned https://github.com/bitcoin/bitcoin/pull/18988#issuecomment-642731876, with [01c31f9](https://github.com/bitcoin/bitcoin/commit/01c31f9dcd15c1009c4b69e820fc0c01b73cd3c9) from branch [ryanofsky@`ipc-export`](https://github.com/ryanofsky/bitcoin/commits/ipc-export), you can connect to an already running `bitcoin-node` through a `<datadir>/sockets/node.sock` socket, which is probably what you would want to do as an external application, rather than starting a bitcoin process.

Let me know what you're trying to do though, since I'm not really sure what information's relevant here.

**Comment by Ayms** [comment]
What I want to do is:

	bitcoin | node-Tor | bitcoin via stdin/stdout or bitcoin <-> node-Tor via unix sockets
			|
		send/receive
		via Tor or
		node-Tor nodes
			|
		optional: node-Tor RDV point
			|
		node-Tor or Tor nodes
			|
		bitcoin node via Tor exit node
			or
		bitcoin <-> node-Tor process

This is one phase of some funding proposals for the more global Convergence concepts, I will let you know when/if it happens and maybe altnet by that time will have evolved, if too complicate to setup another option is to try this with [bitcoin-transactions](https://github.com/Ayms/bitcoin-transactions) which implement a part of the bitcoin protocol, and/or to combine both so the node-Tor node above can be a browser also (WebRTC/WS)

**Comment by ryanofsky** [comment]
> What I want to do is:
> 
> ```
> bitcoin | node-Tor | bitcoin via stdin/stdout or bitcoin <-> node-Tor via unix sockets
```

This is really interesting. Maybe it would be good to create a specific issue to discuss this here or in the node-Tor github repo.

If I were trying to connect bitcoind <--> node-Tor, I think the simplest approach to start with would be to get existing bitcoind `-onion` and `-listenonion` options to work with node-Tor. To do this, no `bitcoind` changes should be needed and node-Tor should just need to support simple subsets of the tor control and socks5 protocols (if it doesn't already). To avoid the need to use plumbing tools like socat, or to modify bitcoind, it would probably be easiest to just start with plain tcp sockets instead of unix sockets or stdin/stdout.

After this worked, I'd start looking into more changes to streamline and package the functionality better:

1. Changing -onion and -listenonion to support unix sockets or pipes instead of tcp
2. Using #18988 (this PR) to write a new altnet driver that can speak socks5/torcontrol to node-Tor. No need to mess with #19160 (multiprocess pr) if the altnet driver is written in c++ or can be part of bitcoin core, but you might use #19160 if you want to write the altnet driver in a different language or more independently of the bitcoin core codebase.
3. Maybe switch from socks5/torcontrol to a custom protocol for bitcoind <--> node-Tor communication. If socks5/torcontrol don't fit well into the altnet framework, maybe do this at the same time as writing the altnet driver.

**Comment by Ayms** [comment]
>This is really interesting. Maybe it would be good to create a specific issue to discuss this here or in the node-Tor github repo.

Thanks, probably @ariard can tell us if this discussion should be moved to other places, I don't find it uninteresting to compare the different approaches

node-Tor does support the socks interface since the very beginning but I don't see the point of using a potentially vulnerable localhost and weak socks protocol for processes that are running on the same machine, we saw not a long time ago that it could lead to security breaches for bitcoin

The Tor project (which is designed for browsing) does not really have the choice today (also for Snowflake), for node-Tor I just don't want to use it, and just don't need it, because browsers for example are not browsing anything but acting as Tor protocol relays/nodes (paradoxically they can't browse anything via WebRTC/WS/xhr interfaces)

So I think that a better approach is pluggable layers with altnet that can be chained including unix sockets and including browser interfaces

What I am proposing can look "ambitious/disruptive or whatever" but I am not far from it and those ideas are not just from yesterday, it's just a matter of time to do it and that's just a part of the Convergence project, 3 with unix sockets looks to be the right option for a universal nodejs driver (not talking about a specific driver for node-Tor, it should work for other nodejs programs), writing it myself... most likely no, I could do it but I can't pretend to be a C/C++ expert for whom it should probably be a piece of cake to do it

**Comment by ariard** [comment]
Still eager to collect feedbacks on the design. Blocker is currently the internal plumbing to communicate across processes. State of the discussion is here : https://github.com/chaincodelabs/libmultiprocess/issues/56

**Pull request closed without merging**